### DISCUSSION TITLE
线上服务能否支持天凤再生牌谱分析

### DISCUSSION THREAD

**Discussion by Preserver6166** [Equim-chan/mjai-reviewer#86] [Ideas]
如mleague的天凤再生牌谱

https://tenhou.net/5/#json={"title":["Mリーグ2019　ファイナルシリーズ 6/23(火)","第2試合　東2-0 供託0点"],"name":["沢崎-228.2","多井+127.9","魚谷+175.5","小林+183.5"],"rule":{"disp":"Mリーグ2019　FINAL 12/12　6/23 第2試合","aka":1},"log":[[[1,0,0],[25000,28900,21100,25000],[24],[31],[11,12,15,19,24,29,32,53,41,42,43,44,47],[12,46,43,25,37,35,22,33,41,18,32,17,14,25,38],[32,15,12,12,11,19,41,60,43,35,29,32,43,47,42],[13,13,14,18,19,27,27,28,31,36,36,39,45],[37,47,31,27,38,25,37,19,15,36,29,26,21,23,34,51],[31,60,60,45,14,13,13,39,60,28,60,18,60,60,19,19],[11,13,14,17,23,27,28,28,33,42,42,46,47],[24,46,34,37,34,36,26,21,22,18,11,32,23,38,43,12],[46,60,11,17,47,42,28,33,21,60,60,60,60,"r42",60],[11,14,15,16,18,21,28,33,35,39,45,46,47],[44,24,17,21,39,43,41,16,12,32,42,19,45,26,22],[21,39,45,60,60,44,60,46,28,35,60,43,16,47,45],["和了",[-1300,-2600,6200,-1300],[2,2,2,"20符4飜1300-2600点","立直(1飜)","門前清自摸和(1飜)","断幺九(1飜)","平和(1飜)"]]]]}

目前是不支持分析的。
以及，再生牌谱目前只能一局一个URL，如果要分析東1~南4的整场比赛，似乎要增加更多的工作量

**Comment by Equim-chan** [comment]
目前有一些担忧的问题所以一直没有做这个功能，之后应该会考虑一下。

另外，你完全可以把一堆局合并到一个 JSON 里，往 `log` 这个列表的最外层加元素就可以了。

**Reply by Preserver6166** [reply]
收到，感谢！

**Reply by Equim-chan** [reply]
今天加上了这个功能。

但是要注意，Mortal 只接受半庄 + 天凤规则，其他规则变种可能出问题。

**Reply by Preserver6166** [reply]
> 今天加上了这个功能。
> 
> 但是要注意，Mortal 只接受半庄 + 天凤规则，其他规则变种可能出问题。

试了下，好像牌谱过长会报bad request，猜测可能是请求包的size有限制？
bad case（去掉最后一局的话可以出结果）

{"title":["19-20赛季 常规赛","2019/09/30（月）第1回战"],"name":["萩原聖人","魚谷侑未","小林剛","白鳥翔"],"rule":{"disp":"2019/09/30（月）第1回战","aka":1},"log":[[[0,0,0],[25000,25000,25000,25000],[25],[41],[16,18,21,27,28,28,29,33,34,37,37,46,46],[35,39,34,46,13,31,45,43,47,33,13,44,16,22,24,36,43,16],[21,28,34,39,60,60,60,60,60,60,"r60",60,60,60,60,60,60,60],[13,13,51,21,23,25,31,32,33,34,36,42,42],[41,29,43,28,22,32,23,14,12,37,15,32,12,41,24,26,38,17],[60,60,60,60,25,60,60,"r13",60,60,60,60,60,60,60,60,60,60],[11,15,17,17,18,21,25,27,28,31,37,42,44],[22,22,39,19,47,45,19,11,24,29,38,16,44,34,44,12,19],[31,11,44,25,60,60,21,28,22,22,15,37,16,44,24,27,38],[12,14,18,23,29,33,53,36,38,39,43,46,47],[14,41,32,17,36,26,27,15,26,23,45,18,45,19,27,42,35],[29,43,41,46,60,23,38,32,12,33,47,15,60,45,60,26,27],["和了",[9700,-7700,0,0],[0,1,0,"40符3飜7700点","立直(1飜)","河底撈魚(1飜)","役牌 發(1飜)"]]],[[0,1,0],[33700,16300,25000,25000],[12],[],[14,21,32,33,35,36,37,37,38,39,47,47,47],[12,11,18,26,26,36,22,27,46,21],[21,60,60,60,60,12,60,60,14,60],[14,51,17,21,24,52,29,31,39,41,41,45,46],[38,22,29,34,43,24,17,42,36,36],[31,29,60,46,60,17,60,21,39,45],[13,15,15,18,24,25,26,31,35,39,41,42,45],[28,28,41,39,23,37,43,47,43,11],[39,31,42,60,18,45,60,60,15,23],[16,16,17,21,27,28,31,32,32,33,33,34,35],[46,24,23,13,18,22,46,42,32,29],[21,46,35,17,60,13,60,60,32],["和了",[-800,-500,-500,1800],[3,3,3,"20符2飜400-700点","門前清自摸和(1飜)","平和(1飜)"]]],[[1,0,0],[32900,15800,24500,26800],[28],[],[15,17,19,21,22,23,27,28,29,33,34,36,47],[15,25,44,24,"c262425",45,41],[36,47,60,19,17,60,60],[14,15,24,26,28,29,31,32,32,33,37,38,47],[27,24,36,53,41,42,39,44],[47,32,"r26",60,60,60,60,60],[13,13,22,24,31,31,33,33,34,39,41,41,42],[21,39,43,29,37,36,23,35],[60,22,42,43,41,41,39,39],[12,16,16,16,17,52,26,27,28,29,39,46,46],[11,51,35,43,26,11,22,13],[39,12,11,60,26,60,60,35],["和了",[3000,0,0,-2000],[0,3,0,"30符2飜2000点","一気通貫(1飜)","ドラ(1飜)"]]],[[2,0,0],[35900,14800,24500,24800],[16],[44],[11,17,22,24,27,28,31,37,38,42,46,47,47],[37,18,34,29,15,23,18,41,32,44,47,14,32,"c393738",16,43,12],[11,31,60,37,60,46,18,60,60,60,42,60,60,47,47,60,60],[12,16,19,24,25,26,26,27,27,35,39,42,43],[46,11,13,46,14,39,34,18,32,24,23,36,14,29,15,13,17],[19,39,43,42,24,60,11,46,60,26,26,"r46",60,60,60,60],[51,17,19,22,24,27,28,31,37,41,42,45,47],[19,43,16,18,38,21,33,31,39,36,"c262728",12,35,29,12,34,43],[31,60,42,47,45,60,60,60,41,39,51,60,19,19,29,12,60],[11,13,14,21,23,25,25,26,28,33,53,37,46],[45,45,33,26,31,22,38,39,17,36,"c511314",35,21,36,21,28,23],[11,46,28,21,23,60,31,26,25,17,38,60,60,39,60,45,45],["和了",[-2000,9000,-4000,-2000],[1,1,1,"満貫2000-4000点","立直(1飜)","門前清自摸和(1飜)","断幺九(1飜)","ドラ(1飜)"]]],[[3,0,0],[33900,22800,20500,22800],[36],[],[11,19,24,29,29,31,32,53,38,39,39,41,44],[15,38,18,43,13,15,45,23,41,21,24,46,45,44,36,22,13,28],[24,15,11,60,60,60,60,60,18,60,60,19,60,31,60,32,46,39],[17,19,24,24,29,31,32,33,38,41,42,42,46],[41,19,14,16,32,39,11,31,36,12,37,"p191919",37,28,38,17,47,16],[24,24,60,29,46,32,60,60,60,60,16,17,60,42,42,60,60,31],[11,11,12,14,15,16,52,28,28,31,33,35,43],[14,34,17,27,18,23,33,36,35,27,"c161415",21,34,34,21,34,37,33],[43,31,11,11,12,14,60,33,60,60,23,60,60,60,60,34,34,60],[12,13,19,22,25,25,26,26,27,32,39,42,45],[51,27,26,44,18,22,22,37,43,13,"p272727",12,14,46,47,42,17,16,43],[42,45,19,60,60,39,32,60,60,12,51,60,60,60,60,60,60,60,60],["流局",[-1500,-1500,1500,1500]]],[[3,1,0],[32400,21300,22000,24300],[12],[],[16,21,24,26,29,34,35,36,39,42,44,47,47],[13,42,12,33,"42p4242",38,39,36,19,18,24,"c111213",18,16],[21,39,16,29,44,60,60,36,36,19,47,47,18,26],[14,17,23,25,26,27,29,32,37,41,41,45,46],[45,45,37,18,25,"41p4141",46,37,46,47,23,11,15,36,"c262527"],[32,37,60,46,14,18,17,60,29,60,60,60,60,60,23],[11,14,14,15,17,17,22,22,22,24,37,39,42],[33,35,44,42,31,29,21,27,28,22,44,16,45,46,38],[11,42,14,60,44,60,60,60,31,33,39,44,15,35,60],[11,12,14,19,21,23,24,28,28,32,34,36,43],[21,38,29,32,41,35,33,43,34,52,44,11,43,26,27],[11,19,43,38,60,21,21,29,32,"r43",60,60,60,60,60],["和了",[0,9300,0,-8300],[1,3,1,"満貫8000点","場風 東(1飜)","役牌 白(1飜)","混一色(2飜)"]]]]}

**Reply by Equim-chan** [reply]
刚刚调整了一下，你可以再试试。

**Reply by Preserver6166** [reply]
> 刚刚调整了一下，你可以再试试。

可以了，感谢！
我这边开发了一个mleague官方牌谱转天凤牌谱的功能，预计会把转换后的牌谱发在B站的专栏space.bilibili.com/395707756
（目前还有些Bug在修，以及官方牌谱本身记错的地方相当多...）。能否授权我将分析结果附在专栏上作为参考？

**Reply by Equim-chan** [reply]
这个没问题的

**Reply by Preserver6166** [reply]
> 这个没问题的

好的。多谢！
再报个bad case...
{"title":["Mリーグ2019　レギュラーシーズン 2/27(木)","第1試合　東2-3 供託0点"],"name":["石橋伸洋(Pr)","魚谷侑未(Px)","松本吉弘(AB)","前原雄大(FC)"],"rule":{"disp":"Mリーグ 2019　2/27 第1試合","aka":1},"log":[[[1,3,0],[-7700,86800,8000,12900],[35],[41],[12,13,51,19,21,22,23,24,25,27,31,45,46],[45,17,16,45,42,11,27,52,44],[19,31,27,46,60,"r25",60,60,60],[14,24,27,27,31,32,32,33,38,39,45,46,47],[36,18,28,15,32,29,17,11,13,16],[46,47,45,39,18,24,27,36,27,60],[11,14,14,17,22,23,23,25,29,29,32,33,43],[47,21,36,29,12,43,24,16,12,34],[11,43,17,47,60,60,"r36",60,60],[15,24,26,26,28,34,35,37,37,38,43,44,46],[37,39,12,33,39,44,35,39,44],[44,43,60,46,60,60,39,60,60],["和了",[-800,-1300,4900,-800],[2,2,2,"30符2飜500-1000点","立直(1飜)","門前清自摸和(1飜)"]]]]}
会报not a four-player game，我看了源代码，好像会报这个错的只有这里https://github.com/Equim-chan/mjai-reviewer/blob/468328fe7e309f5cdf76f34de4f81f832e62c07e/convlog/src/tenhou/log.rs

**Reply by Equim-chan** [reply]
之前提过的。

> 但是要注意，Mortal 只接受半庄 + 天凤规则，其他规则变种可能出问题。

出问题的时候就想一想这种局面是否可能在天凤段位战里出现。这个 case 的问题是出现了负的点数。

**Reply by Preserver6166** [reply]
> 之前提过的。
> 
> > 但是要注意，Mortal 只接受半庄 + 天凤规则，其他规则变种可能出问题。
> 
> 出问题的时候就想一想这种局面是否可能在天凤段位战里出现。这个 case 的问题是出现了负的点数。

噢噢。明白了
